webrender/src/frame.rs

```diff
@@ -1090,16 +1094,26 @@ impl FrameContext {
 
         self.clip_scroll_tree
             .finalize_and_apply_pending_scroll_offsets(old_scrolling_states);
+
+        self.removed_pipelines.extend(scene.removed_pipelines.drain(..));
+
         frame_builder
     }
 
     pub fn update_epoch(&mut self, pipeline_id: PipelineId, epoch: Epoch) {
         self.pipeline_epoch_map.insert(pipeline_id, epoch);
     }
 
-    pub fn make_rendered_document(&self, frame: Frame) -> RenderedDocument {
+    pub fn make_rendered_document(&mut self, frame: Frame) -> RenderedDocument {
         let nodes_bouncing_back = self.clip_scroll_tree.collect_nodes_bouncing_back();
-        RenderedDocument::new(self.pipeline_epoch_map.clone(), nodes_bouncing_back, frame)
+        RenderedDocument::new(
+            PipelineInfo {
+                epochs: self.pipeline_epoch_map.clone(),
+                removed_pipelines: replace(&mut self.removed_pipelines, Vec::new()),
```

Comment:
wouldn't this suffer from the same problem we have with GPU cache, where the consistency of the system depends on strictly producing and consuming rendered documents in order? Instead, I'd like to see the backend being able to produce any number of rendered documents, and Renderer being able to skip any of them, and still be correct. At least the current GPU cache updates are structured this way.

Reply:
On the render backend side I believe I made sure that ordering is preserved and notifications are not lost no matter which order we build the scenes and frames in (that's the reason why the removed pipelines vector goes through all of the various pieces of the render backend and not just the scene and/or frame context).
On the renderer side, I think that it works as long as all documents render to the same swap chain, in the sense that the goal is to notify removed pipelines at least one frame after the last time it was rendered, so even if we skip rendering a document but notify its removed pipelines, said pipelines have already been removed, won't be around next time we do render the document and can be safely removed, I think.

As far as ordering go when sending to the renderer, I think that there shouldn't be any mismatch because the information is passed along with the frame itself (isn't in an separate channel or something that would need extra synchronization).

So I believe that this should work as intended even if we don't render some documents as long as it's in one swap chain (when we support direct composition and friends we'll need to track this information per swap chain). Unless by skipping rendering a document you mean receive a document in the renderer but keep sending drawing the commands of its previous version, in which case we need to add some extra logic to manage that, probably by tracking this information per document. in the renderer.

Reply:
My desire here is to have rendering independent of modelling. The backend should be able to spew as many frames as it needs, and the renderer should be able to consume any subset of those. This is reflected in the type system by the fact we don't mutate most of the things when producing a `RenderedDocument`.

Reply:
I agree with this desire, I don't think this interferes.

Since removed pipelines aren't *states* that define the scene but really *events* associated to it that must be sent once, there is no getting around moving them through the pipeline which requires the events to be mutable.

In addition, we need these events to be ordered with frames that are sent to the renderer. By far the easiest way to preserve this ordering is to send the information along with the frames and require that the renderer (whether or not it ends up rendering the frames) at least collects the events and makes them available to the embedder.

To preserve the non-mutable ref to the scene we would need to track the removed pipelines in parallel to the scene instead of inside the scene, make sure that we associate the right list of removed pipelines with the right scene and not forget to send the removed pipelines of a given scene when we produce a frame out of it. That sounds a lot more error prone. If you care a lot about the scene being passed immutably, then I wouldn't be opposed to using a Cell for the removed pipelines vector.

Reply:
> In addition, we need these events to be ordered with frames that are sent to the renderer.

They don't have to be 1:1. We only have to guarantee that a pipeline is not used after the removed pipeline event arrives. I see the situation exactly matching GPU cache updates, and I want the handling to be consistent. Can we have a single mechanism/entity responsible for those one-time updates? Perhaps, by extending and renaming `UpdateGpuCache` event or something. Just handle it all in the same way.

> That sounds a lot more error prone.

This is much easier than you imply. We already have association of the Document - Scene.